### PR TITLE
bpo-39828: Fix json.tool to catch BrokenPipeError.

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue39828

### CODE REVIEW

**Review by remilapeyre — COMMENTED** [review]
Is it really good to just silence the error? It seems to me that it should at least write something on stderr and set an exit code different than 0.

**Comment by corona10** [comment]
@remilapeyre 

I expect the same behavior as this command.

```
echo "{}" | wc -l | true
```

**Comment by vstinner** [comment]
> Is it really good to just silence the error? It seems to me that it should at least write something on stderr and set an exit code different than 0.

I'm not aware of any tool which logs an error into stderr when stdout pipe is closed, before the producer has time to write all its output into stdout.

**Review by vstinner — APPROVED** [review]
LGTM.

It seems like it's the first time that we start to catch explicitly BrokenPipeError in Python stdlib. Only the subprocess has code to handle its own pipes, which is a more specific case.

I'm not sure if it's the right approach, so I would prefer to get a review from another core dev.

cc @pablogsal @methane @serhiy-storchaka: Does it look like the good design to you to handle BrokenPipeError?

**Comment by remilapeyre** [comment]
> I'm not aware of any tool which logs an error into stderr when stdout pipe is closed, before the producer has time to write all its output into stdout.

At least `wc`, `cat`, `git`, `jq` and `seq` seem to return a non zero code:

```sh
➜  ~ set -o pipefail
➜  ~ echo "{}" | wc -l | true
➜  ~ echo $?
141
➜  ~ echo foo | cat | true
➜  ~ echo $?
141
➜  cpython git:(master) git log | true
➜  cpython git:(master) echo $?
141
➜  ~ echo '{}' | jq '.' | true
➜  ~ echo $?
141
➜  ~ seq 1 10000 | true
➜  ~ echo $?
141
```

and curl sets a non zero exit code and write an error on stderr:

```sh
➜  ~ curl -s https://google.com | true
(23) Failed writing body
➜  ~ echo $?
23
```

This change make `json.tool` behave differently:
```sh
➜  cpython git:(bpo-39828) echo "{}" | ./python.exe -m json.tool | true

➜  cpython git:(bpo-39828) echo $?
0
```
I'm not sure where the empty line comes from. 

I think we should at least set a non zero exit code.

**Comment by vstinner** [comment]
> At least wc, cat, git, jq and seq seem to return a non zero code: (...)

By the way, Python exits with code 120 if it fails to flush stdout or stderr:

```
int
Py_RunMain(void)
{
    int exitcode = 0;
    pymain_run_python(&exitcode);
    if (Py_FinalizeEx() < 0) {
        /* Value unlikely to be confused with a non-error exit status or
           other special meaning */
        exitcode = 120;
    }
    (...)
    return exitcode;
}
```

Change introduced in bpo-5319 by the commit b4ce1fc31be5614d527d77c55018281ebbcd70ab. The issue uses ``> /dev/full`` to test the behavior: interesting test ;-)

**Comment by corona10** [comment]
@remilapeyre  cc @vstinner 

On macOS it returns 0

```
cpython git:(bpo-39828) ✗ echo "{}" | wc -l | true
➜  cpython git:(bpo-39828) ✗ echo $?
0
```

**Comment by corona10** [comment]
@remilapeyre @vstinner 

I've updated the PR.
This solution might be more proper.

```bash
➜  cpython git:(bpo-39828) ✗ echo "{}" | ./python.exe -m json.tool | true
➜  cpython git:(bpo-39828) ✗ echo $?
141
```

**Comment by corona10** [comment]
But this solution is not proper for WIndows.

According to mypy issue, https://github.com/python/mypy/issues/2893
What about just catching the error and firing sys.exit with a constant value.

**Comment by corona10** [comment]
FYI, 7259ef6 returns 

```bash
cpython git:(bpo-39828) ✗ echo "{}" | ./python.exe -m json.tool | true
➜  cpython git:(bpo-39828) ✗ echo $?
32
```

**Comment by remilapeyre** [comment]
> By the way, Python exits with code 120 if it fails to flush stdout or stderr

How, that's interesting. I think it doesn't work here since the pipe is closed when `true` finish so it can't try to flush the extra data. As far as I know there is no way to know if some data is left in the buffer.

> But this solution is not proper for Windows.

@corona10, thanks for looking into this issue, 7259ef675a4682118d5967437591569c27415984 looks good indeed 👍

**Review by remilapeyre — APPROVED** [review]

File: Misc/NEWS.d/next/Library/2020-03-05-00-57-49.bpo-39828.yWq9NJ.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2020-03-05-00-57-49.bpo-39828.yWq9NJ.rst]: The error is not exactly "ignored", but "handled" or "catched". You should rephrase the NEWS entry.

File: Lib/json/tool.py
Comment (vstinner) [inline Lib/json/tool.py]: nitpick: i prefer to avoid single letter variable name, I like "exc" for the except variable.

**Comment by corona10** [comment]
@remilapeyre @vstinner

Thanks for the review :)
I 've updated Victor's review also, Please take look at it :)

**Review by vstinner — APPROVED** [review]
LGTM. Again, I would prefer a second review from another core dev.

**Review by methane — APPROVED** [review]

**Comment by vstinner** [comment]
Is it possible to write a reliable test for BrokenPipeError? Maybe using a string longer than support.test.PIPE_MAX_SIZE to ensure that the pipe is full?

**Comment by corona10** [comment]
@vstinner 

While I tried with this code, the test was not failed...
Is there any wrong with my approach or is there any suggestion to do?

```python
    def test_broken_pipe_error(self):
        cmd = [sys.executable, '-m', 'json.tool']
        large_data = 'xxx' * support.PIPE_MAX_SIZE
        large_data = f'"{large_data}"'
        process = subprocess.run(cmd, input=large_data, capture_output=True, text=True, check=True)
        self.assertEqual(process.stdout.strip(), large_data)
        self.assertEqual(process.stderr, '')
```

**Comment by remilapeyre** [comment]
Did you try using Popen and closing the stdout pipe?

**Comment by corona10** [comment]
> Did you try using Popen and closing the stdout pipe?

Oh It works :) Greeat! Thank you

**Comment by corona10** [comment]
@vstinner Please take a look. :)

File: Lib/test/test_json/test_tool.py
Comment (vstinner) [inline Lib/test/test_json/test_tool.py]: Please avoid hardcoded values. I guess that 32 is errno.EPIPE here.

File: Lib/test/test_json/test_tool.py
Comment (vstinner) [inline Lib/test/test_json/test_tool.py:212]: Where does the ValueError come from?

File: Lib/test/test_json/test_tool.py
Comment (vstinner) [inline Lib/test/test_json/test_tool.py]: I would prefer valid JSON:
```suggestion
        proc.communicate(b'{}')
```

File: Lib/test/test_json/test_tool.py
Comment (vstinner) [inline Lib/test/test_json/test_tool.py:217]: Please add a comment explaining that stdin=PIPE ensures that the stdout close is closed before json.tool attempts to write into stdout.

It's a nice synchronization primite: better than time.sleep() ;-)

File: Lib/test/test_json/test_tool.py
Comment (vstinner) [inline Lib/test/test_json/test_tool.py]: You can remove it. communicate() already closes stdin for you: see Popen._stdin_close().

File: Lib/test/test_json/test_tool.py
Comment (corona10) [reply Lib/test/test_json/test_tool.py]: Oh, cool tip!

File: Lib/test/test_json/test_tool.py
Comment (corona10) [reply Lib/test/test_json/test_tool.py:212]: https://github.com/python/cpython/pull/18779/checks?check_run_id=496883234

```bash
 test_broken_pipe_error (test.test_json.test_tool.TestTool) ... ERROR
  File "D:\a\cpython\cpython\lib\threading.py", line 944, in _bootstrap_inner
    self.run()
  File "D:\a\cpython\cpython\lib\threading.py", line 882, in run
    self._target(*self._args, **self._kwargs)
  File "D:\a\cpython\cpython\lib\subprocess.py", line 1493, in _readerthread
    buffer.append(fh.read())
ValueError: read of closed file
D:\a\cpython\cpython\lib\subprocess.py:1066: ResourceWarning: subprocess 6532 is still running
  _warn("subprocess %s is still running" % self.pid,
ResourceWarning: Enable tracemalloc to get the object allocation traceback
OSError: [Errno 22] Invalid argument

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "D:\a\cpython\cpython\lib\runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "D:\a\cpython\cpython\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "D:\a\cpython\cpython\lib\json\tool.py", line 76, in <module>
    main()
  File "D:\a\cpython\cpython\lib\json\tool.py", line 71, in main
    raise SystemExit(e)
```

**Comment by vstinner** [comment]
@corona10: This fix deserves to be backported to 3.7 and 3.8, but automated backport failed. Can you please try to do the backport manually?

**Comment by corona10** [comment]
Sure :)

File: Lib/test/test_json/test_tool.py
Comment (bl-ue) [inline Lib/test/test_json/test_tool.py:212]: `The test is failed with ValueError on Windows` isn't grammatically correct.  It should be `The test failed with ValueError on Windows` (minus the `is`) 🤔

Comment (vstinner) [reply Lib/test/test_json/test_tool.py:212]: This PR is now closed. You can write a new PR to enhance the skip message.

**Pull request merged**
